pkg/gui/controllers/files_controller.go

```diff
@@ -559,23 +559,46 @@ func (self *FilesController) createStashMenu() error {
 			{
 				DisplayString: self.c.Tr.LcStashAllChanges,
 				OnPress: func() error {
-					return self.handleStashSave(self.git.Stash.Save, self.c.Tr.Actions.StashAllChanges)
+					return self.handleStashSave(self.git.Stash.Save, self.c.Tr.Actions.StashAllChanges, self.c.Tr.NoFilesToStash)
 				},
-				Key: 's',
+				Key: 'a',
+			},
+			{
+				DisplayString: self.c.Tr.LcStashAllChangesKeepIndex,
+				OnPress: func() error {
+					// if there are no staged files it behaves the same as Stash.Save
+					return self.handleStashSave(self.git.Stash.StashAndKeepIndex, self.c.Tr.Actions.StashAllChangesKeepIndex, self.c.Tr.NoFilesToStash)
+				},
+				Key: 'i',
 			},
 			{
 				DisplayString: self.c.Tr.LcStashStagedChanges,
 				OnPress: func() error {
-					return self.handleStashSave(self.git.Stash.SaveStagedChanges, self.c.Tr.Actions.StashStagedChanges)
+					// there must be something in staging otherwise the current implementation mucks the stash up
+					if !self.helpers.WorkingTree.AnyStagedFiles() {
+						return self.c.ErrorMsg(self.c.Tr.NoTrackedStagedFilesStash)
+					}
+					return self.handleStashSave(self.git.Stash.SaveStagedChanges, self.c.Tr.Actions.StashStagedChanges, self.c.Tr.NoTrackedStagedFilesStash)
+				},
+				Key: 's',
+			},
+			{
+				DisplayString: self.c.Tr.LcStashUnstagedChanges,
+				OnPress: func() error {
+					if self.helpers.WorkingTree.AnyStagedFiles() {
+						return self.handleStashSave(self.git.Stash.StashUnstagedChanges, self.c.Tr.Actions.StashUnstagedChanges, self.c.Tr.NoFilesToStash)
+					}
+					// ordinary stash
+					return self.handleStashSave(self.git.Stash.Save, self.c.Tr.Actions.StashUnstagedChanges, self.c.Tr.NoFilesToStash)
 				},
-				Key: 'S',
+				Key: 'u',
```

Comment:
It would be good if we could add a couple of integration tests for the new functionality i.e. a test where we stash staged changes and keep index, and a test where we stash unstaged changes. I can't recall if you've tried your hand at integration tests before but there's a guide at https://github.com/jesseduffield/lazygit/blob/master/docs/Integration_Tests.md

Reply:
Will do! Wanted to iron out all the kinks first!